### PR TITLE
3D diagnostic tendency bug fixes for operational physics

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
Bug fixes to ldiag3d and addition of a qdiag3d flag to turn on tracer tendencies.

### CODE REVIEW

File: physics/GFS_GWD_generic.F90 [hunk 1]
```diff
@@ -91,10 +92,13 @@ subroutine GFS_GWD_generic_pre_run(                               &
       endif   ! end if_nmtvr
 
       if (lssav) then
-        if (ldiag3d) then
+        if (ldiag3d .and. gwd_generic_tend) then
+          write(0,*) 'gwd_generic_tend'
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.F90]: We should probably remove this write statement

File: physics/GFS_GWD_generic.F90 [hunk 2]
```diff
@@ -150,7 +154,8 @@ subroutine GFS_GWD_generic_post_run(lssav, ldiag3d, dtf, dusfcg, dvsfcg, dudt, d
         dugwd(:) = dugwd(:) + dusfcg(:)*dtf
         dvgwd(:) = dvgwd(:) + dvsfcg(:)*dtf
 
-        if (ldiag3d) then
+        if (ldiag3d .and. gwd_generic_tend) then
+          write(0,*) 'gwd_generic_tend'
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.F90]: Dito

File: physics/GFS_GWD_generic.meta [hunk 3]
```diff
@@ -118,6 +118,20 @@
   type = logical
   intent = in
   optional = F
+[dudt]
+  standard_name = tendency_of_x_wind_due_to_model_physics
+  long_name = zonal wind tendency due to model physics
+  units = m s-2
+  dimensions = (horizontal_dimension,vertical_dimension)
+  type = real
+  kind = kind_phys
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.meta:127]: I am missing the intent and optional flags for dudt and dvdt

File: physics/GFS_GWD_generic.meta [hunk 4]
```diff
@@ -127,6 +141,20 @@
   kind = kind_phys
   intent = in
   optional = F
+[du3dt]
+  standard_name = cumulative_change_in_x_wind_due_to_orographic_gravity_wave_drag
+  long_name = cumulative change in x wind due to orographic gravity wave drag
+  units = m s-1
+  dimensions = (horizontal_dimension,vertical_dimension)
+  type = real
+  kind = kind_phys
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.meta:152]: As above, but for du3dt and dv3dt

File: physics/GFS_GWD_generic.meta [hunk 5]
```diff
@@ -145,6 +173,12 @@
   kind = kind_phys
   intent = in
   optional = F
+[gwd_generic_tend]
+  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
+  long_name = true if GFS_GWD_generic should calculate tendencies
+  units = flag
+  dimensions = ()
+  type = logical
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.meta:185]: dito

File: physics/GFS_GWD_generic.meta [hunk 6]
```diff
@@ -287,6 +321,12 @@
   kind = kind_phys
   intent = inout
   optional = F
+[gwd_generic_tend]
+  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
+  long_name = true if GFS_GWD_generic should calculate tendencies
+  units = flag
+  dimensions = ()
+  type = logical
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.meta:334]: dito

File: physics/GFS_MP_generic.F90 [hunk 7]
```diff
@@ -42,12 +42,14 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
         do k=1,levs
           do i=1,im
             save_t(i,k) = gt0(i,k)
-            save_q(1:im,:,1) = gq0(1:im,:,1)
```
Comment (DomHeinzeller) [inline physics/GFS_MP_generic.F90:45]: this assignment of the entire column inside the k loop didn't make any sense, thanks for fixing it

File: physics/cires_ugwp.meta [hunk 8]
```diff
@@ -850,6 +850,60 @@
   type = integer
   intent = in
   optional = F
+[ldu3dt_ogw]
+  standard_name = cumulative_change_in_x_wind_due_to_orographic_gravity_wave_drag
+  long_name = cumulative change in x wind due to orographic gravity wave drag
+  units = m s-1
+  dimensions = (horizontal_dimension,vertical_dimension)
+  type = real
+  kind = kind_phys
```
Comment (DomHeinzeller) [inline physics/cires_ugwp.meta:859]: Please add intent and optional attributes for all new variables.

File: physics/gwdc.f [hunk 9]
```diff
@@ -1498,13 +1498,14 @@ subroutine gwdc_post_run(                                         &
...
         dvgwd(:) = dvgwd(:) + taucty(:)*dtf
-
-        if (ldiag3d) then
-          du3dt(:,:) = du3dt(:,:) + gwdcu(:,:)  * dtf
-          dv3dt(:,:) = dv3dt(:,:) + gwdcv(:,:)  * dtf
-        endif
       endif   ! end if_lssav
 
+      if (ldiag3d) then
+         write(0,*) 'update gwdc tend'
```
Comment (DomHeinzeller) [inline physics/gwdc.f]: Please remove the write statement

**Review by DomHeinzeller — CHANGES_REQUESTED** [review]
This is great, thanks very much. I left a few notes on code to clean up. I would also like to have @grantfirl take a look.

**Comment by SamuelTrahanNOAA** [comment]
The gwd_generic_tend functionality is separate from the ldiag3d/qdiag3d
tendencies.  It looks like it was a half-implemented attempt at splitting
the gravity wave drag tendencies out to more terms.  I'm not sure it
works.  The two ldiag3d gravity wave drag tendencies from the days of yore
do work.

**Comment by SamuelTrahanNOAA** [comment]
(Prior, incorrect, comment was deleted.)

The gwd_generic_tend is a flag I added to deal with inconsistencies in the
logic of GWD tendencies.  There is a second set of GWD tendencies that I
think Valary Yudin added, and I'm not sure those work.  I've left them
alone as best I can and got the two old GWD tendencies working.

File: physics/GFS_MP_generic.F90 (hunk 7 above)
Comment (SamuelTrahanNOAA) [reply physics/GFS_MP_generic.F90:45]: There is an error here.  The save_qv variable should be used instead.

**Comment by SamuelTrahanNOAA** [comment]
I think all issues are now resolved.

I'm still not sure what to do about the duplicated GWD and PBL diagnostics variables, but I suspect that problem is much more widespread than just Valery's additions.

**Comment by SamuelTrahanNOAA** [comment]
I just fixed additional bugs.  The code that was adding the PBL tendencies subtracted those same tendencies from the OGW tendency variables.  I had fixed that in one place, but not another.  The recent batch of CCPP updates added many new bugs to the diagnostic tendencies.  This time I think I've fixed the ones in the operational physics.

**Comment by DomHeinzeller** [comment]
@SamuelTrahanNOAA just to confirm - is this PR now ok from your side?

**Comment by SamuelTrahanNOAA** [comment]
Yes.

File: physics/cires_ugwp.F90 [hunk 10]
```diff
@@ -172,6 +174,12 @@ subroutine cires_ugwp_run(do_ugwp, me,  master, im,  levs, ntrac, dtp, kdt, lonr
     real(kind=kind_phys),    intent(out), dimension(im)      :: tau_mtb, tau_ogw, tau_tofd, tau_ngw
     real(kind=kind_phys),    intent(out), dimension(im, levs):: gw_dudt, gw_dvdt, gw_dtdt, gw_kdis
     real(kind=kind_phys),    intent(out), dimension(im, levs):: dudt_mtb, dudt_ogw, dudt_tms
+
+    ! These arrays are only allocated if ldiag=.true.
+    real(kind=kind_phys),    intent(inout), dimension(im, levs) :: ldu3dt_ogw, ldv3dt_ogw, ldt3dt_ogw
```
Comment (DomHeinzeller) [inline physics/cires_ugwp.F90]: If these arrays are only allocated if ldiag3d (not ldiag) is true, then they should be assumed-size arrays (:,:) instead of having explicit dimensions.

File: physics/GFS_GWD_generic.meta [hunk 11]
```diff
@@ -145,6 +177,13 @@
   kind = kind_phys
   intent = in
   optional = F
+[gwd_generic_tend]
+  standard_name = true_if_GFS_GWD_generic_should_calculate_tendencies
```
Comment (DomHeinzeller) [inline physics/GFS_GWD_generic.meta:181]: I believe the standard name should be something like flag_for_... @grantfirl once again your opinion would be appreciated.

**Review by DomHeinzeller — CHANGES_REQUESTED** [review]
A few minor comments. I would also get @grantfirl's opinion on moving the tendency calculation inside physics schemes, in particular moninedmf.f (Grant has been working much more on/with the various PBL schemes than I have).

**Comment by SamuelTrahanNOAA** [comment]
The array dimension and flag name changes need to be made.  I'll work on
that.

Moving the tendency calculations into the physics schemes is necessary
since the generic interstitial code doesn't know which tendencies should
come out of a scheme.  That information is scheme-specific.

File: physics/GFS_DCNV_generic.meta [hunk 12]
```diff
@@ -25,6 +25,14 @@
   type = logical
   intent = in
   optional = F
+[qdiag3d]
+  standard_name = flag_tracer_diagnostics_3D
```
Comment (grantfirl) [inline physics/GFS_DCNV_generic.meta:29]: Isn't it just specific humidity tendencies instead of all tracers?

File: physics/GFS_GWD_generic.meta (hunk 11 above)
Comment (grantfirl) [reply physics/GFS_GWD_generic.meta:181]: Agree: flag_for_generic_gwd_tendencies is better, although "generic" is a bit vague in this context. You mean generic as opposed to tendencies that Valery put in for the CIRES UGWP scheme?

File: physics/GFS_MP_generic.F90 [hunk 13]
```diff
@@ -42,12 +42,24 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
...
+           do k=1,levs
+              do i=1,im
+                 ! Here, gq0(...,1) is used instead of gq0_water_vapor
+                 ! to be consistent with the GFS_MP_generic_post_run
+                 ! code.
+                 save_qv(i,k) = gq0(i,k,1)
+              enddo
+           enddo
+        endif
+        if(do_aw) then
```
Comment (grantfirl) [inline physics/GFS_MP_generic.F90:57]: I'm curious why this special treatment for do_aw...

File: physics/GFS_MP_generic.F90 [hunk 14]
```diff
@@ -42,12 +42,24 @@ subroutine GFS_MP_generic_pre_run(im, levs, ldiag3d, do_aw, ntcw, nncl, ntrac, g
...
-        do n=ntcw,ntcw+nncl-1
-          save_q(1:im,:,n) = gq0(1:im,:,n)
-        enddo
+        if(qdiag3d) then
+           do k=1,levs
+              do i=1,im
+                 ! Here, gq0(...,1) is used instead of gq0_water_vapor
+                 ! to be consistent with the GFS_MP_generic_post_run
+                 ! code.
+                 save_qv(i,k) = gq0(i,k,1)
```
Comment (grantfirl) [inline physics/GFS_MP_generic.F90:53]: All things equal, I'd prefer to rely on passing in gq0_water_vapor instead of the whole array and hard-coding the water vapor index. I see that the whole array is used below for do_aw, in which case, OK, no need to pass in the data twice.

File: physics/GFS_MP_generic.F90 [hunk 15]
```diff
@@ -256,9 +269,15 @@ subroutine GFS_MP_generic_post_run(im, ix, levs, kdt, nrcm, ncld, nncl, ntcw, nt
           do k=1,levs
             do i=1,im
               dt3dt(i,k) = dt3dt(i,k) + (gt0(i,k)-save_t(i,k)) * frain
-!              dq3dt(i,k) = dq3dt(i,k) + (gq0(i,k,1)-save_qv(i,k)) * frain
             enddo
           enddo
+          if (qdiag3d) then
+             do k=1,levs
+                do i=1,im
+                   dq3dt(i,k) = dq3dt(i,k) + (gq0(i,k,1)-save_qv(i,k)) * frain
```
Comment (grantfirl) [inline physics/GFS_MP_generic.F90:291]: similar comment to above. I'd prefer to see just the water vapor slice passed in.

File: physics/moninedmf.f [hunk 16]
```diff
@@ -1037,6 +1041,17 @@ subroutine hedmf_run (ix,im,km,ntrac,ntcw,dv,du,tau,rtg,          &
             rtg(i,k,1) = rtg(i,k,1)+qtend
             dtsfc(i)   = dtsfc(i)+cont*del(i,k)*ttend
             dqsfc(i)   = dqsfc(i)+conq*del(i,k)*qtend
+            if(lssav .and. ldiag3d) then
```
Comment (grantfirl) [inline physics/moninedmf.f:1045]: I'm a little confused why tendencies specific to moninedmf are needed. Shouldn't the ones calculated in GFS_PBL_generic suffice? If not, would you want to add ".and. .not. pbl_generic_tend"?

File: physics/GFS_MP_generic.F90 (hunk 14 above)
Comment (climbfuji) [reply physics/GFS_MP_generic.F90:53]: Instead of `gq0(i,k,1)` we should be using `gq0(i,k,ntqv)` in this case.

**Comment by grantfirl** [comment]
I'm curious what the strategy for tendencies in general is. Process-split schemes like surface layer, PBL, and GWD must output a tendency while time-split schemes like the others don't necessarily have to except for diagnostic purposes, although they are certainly good diagnostics to have. My opinion (not that anyone asked for it) would be to basically treat all CCPP schemes like process-split schemes and to calculate tendencies by default automatically through the CCPP framework and to not allow schemes to modify the state internally. For time-split schemes, those tendencies could then simply be applied automatically after they're called too.

With the exception of CIRES UGWP, we've been putting all diagnostic tendencies in the "scheme-generic" interstitials because presumably we'd like to have tendencies calculated in the same way for any scheme of a given type. IMO, relying on all schemes to calculate their own tendencies internally means a lot of repeated code and potential confusion...

File: physics/GFS_MP_generic.F90 (hunk 14 above)
Comment (grantfirl) [reply physics/GFS_MP_generic.F90:53]: Thanks @climbfuji. I couldn't remember if there was a named index for water vapor or not. Since there is, as long as the whole tracer array is needed for do_aw, using the named index would definitely be the preferred solution.

**Comment by SamuelTrahanNOAA** [comment]
No.  Ozone is in the dq3dt array.

**Comment by SamuelTrahanNOAA** [comment]
Generic refers to the GFS generic interstitial "schemes."

**Comment by SamuelTrahanNOAA** [comment]
Same here.  When I found code I didn't understand, I tried to leave the old
behavior as closely as possible.

**Comment by SamuelTrahanNOAA** [comment]
Passing in both gq0 and gq0_water_vapor will break assumptions in the
compiler's optimizer.  The resulting executable may not work.

**Comment by SamuelTrahanNOAA** [comment]
The pbl_generic_tend flag is missing.

**Comment by SamuelTrahanNOAA** [comment]
Some schemes output their tendencies instead of modifying the state.  In
that situation, the diagnostic tendencies are calculated from the scheme's
output tendencies.  That is the ideal solution from a framework standpoint.

You cannot calculate the tendencies from state variables in the generic
interstitials because the generic interstitials don't know which tendencies
are provided by the scheme.  For example, you may have a PBL scheme that
also generates OGW tendencies.  In that case the interstitial has no idea
how to split the two.

Most importantly, schemes will want to split the generic tendencies (like
"convective gravity wave drag") into more logical components.  That is not
something that can be done in a generic interstitial because the generic
interstitial has no idea what the internal processes are within the scheme.

In other words, there are only two options:

1. A scheme outputs its tendency for that timestep instead of modifying the
state.  The interstitial accumulates those tendencies to make diagnostic
accumulated tendencies.

2. A scheme modifies the state directly, and also outputs the diagnostic
accumulated tendencies.

The current approach of modifying the state, and having the interstitial do
the math, does not work (except perhaps for microphysics).

Sincerely,
Sam Trahan

File: physics/GFS_GWD_generic.meta (hunk 11 above)
Comment (SamuelTrahanNOAA) [reply physics/GFS_GWD_generic.meta:181]: GFS_GWD_generic is the name of the scheme that calculates the tendencies if that flag is true

File: physics/GFS_DCNV_generic.meta (hunk 12 above)
Comment (SamuelTrahanNOAA) [reply physics/GFS_DCNV_generic.meta:29]: No.  It includes ozone tendencies.

File: physics/GFS_MP_generic.F90 (hunk 13 above)
Comment (SamuelTrahanNOAA) [reply physics/GFS_MP_generic.F90:57]: I don't know what do_aw is for, so I left its implementation unmodified as much as possible.

File: physics/moninedmf.f (hunk 16 above)
Comment (SamuelTrahanNOAA) [reply physics/moninedmf.f:1045]: The pbl_generic_tend should be in there.  I will fix that.

**Review by SamuelTrahanNOAA — COMMENTED** [review]
> I'm curious what the strategy for tendencies in general is. Process-split schemes like surface layer, PBL, and GWD must output a tendency while time-split schemes like the others don't necessarily have to except for diagnostic purposes, although they are certainly good diagnostics to have. My opinion (not that anyone asked for it) would be to basically treat all CCPP schemes like process-split schemes and to calculate tendencies by default automatically through the CCPP framework and to not allow schemes to modify the state internally. For time-split schemes, those tendencies could then simply be applied automatically after they're called too.
> 
> With the exception of CIRES UGWP, we've been putting all diagnostic tendencies in the "scheme-generic" interstitials because presumably we'd like to have tendencies calculated in the same way for any scheme of a given type. IMO, relying on all schemes to calculate their own tendencies internally means a lot of repeated code and potential confusion...

Some schemes output their tendencies instead of modifying the state.  In that situation, the diagnostic tendencies are calculated from the scheme's output tendencies.  That is the ideal solution from a framework standpoint.

You cannot calculate the tendencies from state variables in the generic interstitials because the generic interstitials don't know which tendencies are provided by the scheme.  For example, you may have a PBL scheme that also generates OGW tendencies.  In that case the interstitial has no idea how to split the two.

Most importantly, schemes will want to split the generic tendencies (like "convective gravity wave drag") into more logical components.  That is not something that can be done in a generic interstitial because the generic interstitial has no idea what the internal processes are within the scheme.

In other words, there are only two options:

1. A scheme outputs its tendency for that timestep instead of modifying the state.  The interstitial accumulates those tendencies to make diagnostic accumulated tendencies.

2. A scheme modifies the state directly, and also outputs the diagnostic accumulated tendencies.

The current approach of modifying the state, and having the interstitial do the math, does not work (except perhaps for microphysics).

**Comment by SamuelTrahanNOAA** [comment]
This pull request will be replaced.
